download/download.go

```diff
@@ -17,23 +26,155 @@ Download: Downloads files from the Sensitive Data Archive (SDA). If a directory
+		err = os.MkdirAll(filePath, os.ModePerm)
+		if err != nil {
+			return fileName, err
+		}
+	}
+
+	return fileName, nil
+}
+
+// Downloads a file from the url to the filePath location
+func downloadListFile(url string, filePath string) error {
+
+	// Get the file from the provided url
+	resp, err := http.Get(url)
+	if err != nil {
+		return fmt.Errorf("failed to download file, reason: %v", err)
+	}
+	defer resp.Body.Close()
+
+	// Create the file in the current location
+	out, err := os.Create(filePath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+
+	// Write the body to file
+	_, err = io.Copy(out, resp.Body)
+	defer out.Close()
+
+	return err
+
+}
+
+// Read the urls_list.txt file and return the urls of the files in a list
+func getFilesUrls(urlsFilePath string) (urlsList []string, err error) {
+
+	urlsFile, err := os.Open(filepath.Clean(urlsFilePath))
+	if err != nil {
+		return nil, err
+	}
+	defer urlsFile.Close()
+
+	scanner := bufio.NewScanner(urlsFile)
+	for scanner.Scan() {
+		urlsList = append(urlsList, scanner.Text())
+	}
+
+	if len(urlsList) == 0 {
+		return urlsList, fmt.Errorf("failed to get list of files, empty file")
+	}
+
+	return urlsList, scanner.Err()
+}
+
+// Download function downaloads the files included in the urls_list.txt file.
+// The argument can be a local file or a url to an S3 folder
+func Download(args []string) error {
 	// Parse flags. There are no flags at the moment, but in case some are added
 	// we check for them.
 	err := Args.Parse(os.Args[1:])
 	if err != nil {
-		log.Fatalf("Argument parsing failed, reason: %v", err)
+		return fmt.Errorf("failed parsing arguments, reason: %v", err)
 	}
 
 	// Args() returns the non-flag arguments, which we assume are filenames.
 	urls := Args.Args()
+	if len(urls) == 0 {
+		return fmt.Errorf("failed to find location of files, no argument passed")
+	}
+
+	var currentPath, urlsFilePath string
+	currentPath, err = os.Getwd()
+	if err != nil {
+		return fmt.Errorf("failed to get current path, reason: %v", err)
+	}
+
+	switch {
+	// Case where the uses passes the url to the s3 folder where the data exists
+	// Download the urls_list.txt file first and then the data files
+	// e.g. https://some/url/to/folder/
+	case strings.HasSuffix(urls[0], "/") && regexp.MustCompile(`https?://`).MatchString(urls[0]):
+		urlsFilePath = currentPath + "/urls_list.txt"
+		err = downloadListFile(urls[0]+"urls_list.txt", urlsFilePath)
+		if err != nil {
+			return err
+		}
+	// Case where the user passes the url directly to urls_list.txt
+	// e.g. https://some/url/to/urls_list.txt
+	case regexp.MustCompile(`https?://`).MatchString(urls[0]):
+		urlsFilePath = currentPath + "/urls_list.txt"
+		log.Info(urls[0], urlsFilePath)
+		err = downloadListFile(urls[0], urlsFilePath)
+		if err != nil {
+			return err
+		}
+	// Case where the user passes a file containg the urls to download
+	// e.g. /some/folder/to/file.txt
+	default:
+		urlsFilePath = urls[0]
+	}
```

Comment:
No case where the user can supply the URL to a single datafile ?
we can not assume that the user will always download all files every time.

Reply:
That's covered but leaving only the wanted files in the `urls_list.txt` file. Shouldn't that be enough?